### PR TITLE
[branch] Work in progress - commit alpha version of webcam recording

### PR DESCRIPTION
**Pull request by kesiena115**
This branch contains the first trial implementation of webcam recording. The URL for this feature is `/teachermoments/valpha`

It is not fully implemented because it encounters an error in `server/endpoints/audio.js` line 25. The error message is 'Received undefined bytes of audio'. This is because the video data (i.e., webcam recording) is not being sent to the server. I've not been able to figure out what's wrong here.

The following screenshots illustrate the webcam recording, console.log error messages on the browser, and the stack trace of the server-side error.

## Screenshots
screenshot 2017-06-08 11 33 26
screenshot 2017-06-08 04 34 47

### CODE REVIEW

File: ui/src/message_popup/playtest/video_alpha_experience_page.jsx [hunk 1]
```diff
@@ -0,0 +1,263 @@
+/* @flow weak */
+import _ from 'lodash';
+import React from 'react';
+import uuid from 'uuid';
+import VelocityTransitionGroup from "velocity-react/velocity-transition-group";
+
```
Comment (kesiena115) [inline ui/src/message_popup/playtest/video_alpha_experience_page.jsx]: This is pretty much a copy of `danson_experience_page.jsx` (i.e., the version that existed about  a month ago). I simply added webcam recording at the top of the intro page.

File: ui/src/message_popup/playtest/video_alpha_experience_page.jsx [hunk 2]
```diff
@@ -0,0 +1,263 @@
...
+      getNextQuestion={this.onGetNextQuestion}
+      summaryEl={this.renderSummaryEl}
+      onLogMessage={this.onLogMessage}
+    />;
+  },
+
+  renderIntro() {
+    return (
+      <IntroWithEmail defaultEmail={this.state.email} onDone={this.onStart}>
+        <VideRecorder />
```
Comment (kesiena115) [inline ui/src/message_popup/playtest/video_alpha_experience_page.jsx]: We just added the VideoRecorder here to see whether it works

File: ui/src/message_popup/playtest/video_alpha_experience_page.jsx [hunk 3]
```diff
@@ -0,0 +1,284 @@
...
+      getNextQuestion={this.onGetNextQuestion}
+      summaryEl={this.renderSummaryEl}
+      onLogMessage={this.onLogMessage}
+    />;
+  },
+
+  renderIntro() {
+    return (
+      <IntroWithEmail defaultEmail={this.state.email} onDone={this.onStart}>
+        <VideoRecorder />
```
Comment (kesiena115) [inline ui/src/message_popup/playtest/video_alpha_experience_page.jsx]: This isn't where we'd use this normally. Added it here in the intro page just to see whether it works.

File: ui/src/components/video_recorder.jsx [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+    window.ConcatenateBlobs(this.blobs, this.blobs[0].type, this.onBlobsDone);
+  },
+
+  onMediaSuccess(stream) {
+    this.mediaRecorder = new MediaStreamRecorder(stream); 
+    this.mediaRecorder.mimeType = 'video/webm';
+    this.mediaRecorder.ondataavailable = this.onDataAvailable;
+    this.mediaRecorder.start(1000);
+    this.stream = stream;
+    this.setState({videoURL: URL.createObjectURL(stream)});
```
Comment (kevinrobinson) [inline ui/src/components/video_recorder.jsx]: This sets the URL so we can show the video in the browser as it's recording.

File: ui/src/components/video_recorder.jsx [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+
+  onDoneUploading(event) {
+    console.log("Done uploading");
+  },
+
+  onStopRecording() {
+    this.mediaRecorder.stop();
+    this.stream.stop();
+    // this.mediaRecorder.save();
+    window.ConcatenateBlobs(this.blobs, this.blobs[0].type, this.onBlobsDone);
```
Comment (kevinrobinson) [inline ui/src/components/video_recorder.jsx]: This is injected by importing https://github.com/streamproc/MediaStreamRecorder

File: ui/src/components/video_recorder.jsx [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+
+    console.log("onBlobsDone");
+    console.log(megaBlob);
+    // var blobURL = URL.createObjectURL(megaBlob);
+    // this.uploadBlob(this.blobs[0]);
+    // this.setState({ blobURL });
+  },
+
+  onDataAvailable(blob) {
+    // POST/PUT "Blob" using FormData/XHR2
```
Comment (kevinrobinson) [inline ui/src/components/video_recorder.jsx]: This is where we were stuck - we couldn't get data out of the video element and sent up to the server.  We tried a few different things.

We initially tried to store all the blobs, concatenate them in the browser, and push them up.  Getting the bytes for blobs that were streamed worked - we put those into `this.blobs` and they had realistic byte sizes.  But no bytes got sent over the wire.

To debug, we tried to upload a single blob using FormData/XHR2, and that seemed to work from the browser perspective, but still no payload went over the wire to the server.  To debug, we tried creating a blobURL from the blob and see if we could open the local video in the local browser, which worked.  The blob also had a realistic number of bytes.

That's where we got stuck and stopped for now.

**Comment by kevinrobinson** [comment]
@kesiena115 Thanks for all your work pairing and trying to prototype this!  Merging it into a feature branch in the main repo, so we can pick it up again later this summer or work on this with a future student.

**Pull request merged**